bee-vote/src/context.rs

```diff
@@ -0,0 +1,256 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Context information for the voting pool.
+
+use crate::{Error, Opinion, Opinions};
+
+use bee_message::prelude::{MessageId, TransactionId};
+
+use std::fmt;
+
+/// Initial "liked" value for a new `Context`.
+pub const LIKED_INITIAL: f64 = -1.0;
+
+/// Object type of a vote.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum VoteObject {
+    /// Conflict type and related transaction ID.
+    Conflict(TransactionId),
+    /// Timestamp type and related message ID.
+    Timestamp(MessageId),
+}
+
+impl fmt::Display for VoteObject {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            Self::Conflict(id) => write!(f, "Conflict({})", id),
+            Self::Timestamp(id) => write!(f, "Timestamp({})", id),
+        }
+    }
+}
+
+impl VoteObject {
+    /// Get a `MessageId` if this is a Timestamp object.
+    /// Returns `Some(id)` if this is the case, `None` if not.
+    pub fn message_id(&self) -> Option<MessageId> {
+        if let Self::Timestamp(id) = *self {
+            Some(id)
+        } else {
+            None
+        }
+    }
+
+    /// Get a `TransactionId` if this is a Conflict object.
+    /// Returns `Some(id)` if this is the canse, `None` if not.
+    pub fn transaction_id(&self) -> Option<TransactionId> {
+        if let Self::Conflict(id) = *self {
+            Some(id)
+        } else {
+            None
+        }
+    }
+}
+
+/// Builder pattern struct for instantiating a `VoteContext`.
+pub struct VoteContextBuilder {
+    /// Object of the voting (conflict or timestamp), and associated ID.
+    object: VoteObject,
+    /// Opinions held by this context on the vote object.
+    opinions: Option<Opinions>,
```

Comment:
Hi, is there a difference between `opinions` being `None` and `opinions` being `Some(Opinions(vec![])`?

Reply:
No – it gives a nice way to produce an error on building with no opinions through ok_or, but there is no other functional difference. We could definitely achieve the same thing just through an empty check.

Reply:
IMHO I'd avoid the `Option` to have less indirection. But I leave it to your judgement :)

Reply:
The one reason I can think of to use `Option<Opinions>` is if you wanted to differentiate errors for "opinions never set" (programmer error since `with_initial_opinion(s)` was never called on the builder), or "opinions set empty" (logic/node error? since an empty opinion set was generated when it should be non-empty).

I personally lean towards having more descriptive errors when possible, but no strong opinion here.

Reply:
I've removed the `Option`, since it feels a bit more ergonomic that way.